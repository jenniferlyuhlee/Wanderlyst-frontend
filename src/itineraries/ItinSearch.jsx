import React, { useState } from "react";
import WanderlystApi from "../utils/api";
import cleanData from "../helpers/cleanData";
import ItinCard from "./ItinCard";

function ItinSearch( ){
    const initialState = {
        title: "",
        country: "",
        duration: "",
        tags: []
    }
    const [itineraries, setItineraries] = useState(null);
    const [searchQuery, setSearchQuery] = useState(initialState);

    // general callback to update targeted form data (text inputs)
    function handleInputChange(e){
        const { name, value } = e.target
        setSearchQuery(data => ({...data, [name]: value}));
    }
    // general callback to update targeted form data (select input)
    function handleSelectChange(e){
        const { options } = e.target;
        const tags = Array.from(options)
                                    .filter(option => option.selected)
                                    .map(option => option.value);
        setSearchQuery(data => ({...data, tags}));

    }

    // handler for form submission, sends cleaned search query to API
    // displays corresponding itineraries
    async function handleSubmit(e) {
        e.preventDefault();
        const cleanedQuery = cleanData(searchQuery);
        setItineraries(await WanderlystApi.getAllItins(cleanedQuery))
    }

    return(
        <div className = "container px-5 mb-5 body-cont">
            <h3 className = "text-center pb-3">
                Where will you go next?
            </h3>
            <form onSubmit = {handleSubmit} className = "row">
                <div className = "col-12 form-floating p-0">
                    <input id = "title"
                        name = "title"
                        className = "form-control"
                        placeholder = "title"
                        value = { searchQuery.title }
                        onChange = {handleInputChange} />
                    <label htmlFor = "title">Title</label>
                </div>
                <div className = "col-6 form-floating p-0">
                    <input id = "country"
                        name = "country"
                        className = "form-control"
                        placeholder = "country"
                        value = { searchQuery.country }
                        onChange = {handleInputChange} />
                    <label htmlFor = "country">Country</label>
                </div>
                <div className ="col-6 form-floating p-0">
                    <input id = "duration"
                            name = "duration"
                            className = "form-control"
                            type = "number"
                            min = "1"
                            placeholder = "duration"
                            value = { searchQuery.duration }
                            onChange = {handleInputChange} />
                    <label htmlFor = "duration">Max duration (days)</label>
                </div>
                <div className="col-12 form-floating p-0">
                    <select multiple 
                        size = "3"
                        id="tags"
                        name = "tags"
                        className="form-select" 
                        value={ searchQuery.tags}
                        onChange = {handleSelectChange} 
                        style = {{height: '5rem'}}
                    >
                        <option value="Foodie">Foodie</option>
                        <option value="Explorer">Explorer</option>
                        <option value="Shopaholic">Shopaholic</option>
                        <option value="Naturalist">Naturalist</option>
                        <option value="Daredevil">Daredevil</option>
                        <option value="Flâneur">Flâneur</option>
                        <option value="Reveller">Reveller</option>
                        <option value="Aesthete">Aesthete</option>
                        <option value="Relaxationist">Relaxationist</option>
                        <option value="Photographer">Photographer</option>
                    </select>
                    <label htmlFor="tag">Filter by tags</label>
                </div>
                <button className = "btn rounded-pill btn-lg btn-primary mt-3">
                    Search
                </button>
            </form>
        {itineraries ? 
            <div className = "mt-3">
                <i>Showing {`${itineraries.length}
                ${itineraries.length === 1 ? "result" : "results"}:`}</i>
                {itineraries.map(i => (
                    <ItinCard key={i.id} itinerary = {i}/>
                ))}
            </div>
            :
            null}
        </div>
    )
}

export default ItinSearch;
