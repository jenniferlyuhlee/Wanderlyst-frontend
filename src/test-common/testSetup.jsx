import React, { useState } from "react";
import UserContext from "../auth-user/UserContext";
import WanderlystApi from "../utils/api";
import '@testing-library/jest-dom';

const demoUser = {
  username: "testadmin", 
  firstName: "testfirst",
  lastName: "testlast",
  location: "testloc",
  bio: "testbio",
  profilePic: null,
  isAdmin: true,
};

const demoTags = [
  {id: 1, name: 'Foodie', description: 'Savor the world one bite at a time. Dive into a journey packed with delicious local cuisines, hidden culinary gems, and must-try dishes that will leave your taste buds tingling.'},
  {id: 2, name: 'Explorer', description: 'Step off the beaten path and uncover the unknown. From ancient ruins to vibrant cityscapes, let your curiosity guide you through unforgettable experiences.'}
]

const UserProvider = ({ children, 
                        currUser = demoUser, 
                        setCurrUser, 
                        tags = demoTags,
                        logout,
                        hasLikedItin,
                        toggleLike }) => {
  const [user, setUser] = useState(currUser);

  return (
    <UserContext.Provider value={{ currUser: user, 
                                  setCurrUser: setUser, 
                                  tags, logout, 
                                  hasLikedItin, toggleLike}}>
      {children}
    </UserContext.Provider>
  );
};

export { UserProvider };