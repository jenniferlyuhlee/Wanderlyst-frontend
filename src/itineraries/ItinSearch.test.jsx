import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { UserProvider } from "../test-common/testSetup";
import { MemoryRouter} from "react-router-dom";
import WanderlystApi from "../utils/api";
import ItinSearch from "./ItinSearch";
import { expect, describe, test} from "vitest";

describe("ItinSearch component", () => {
    // Mock the API methods before running tests
    beforeEach(() => {
        vi.spyOn(WanderlystApi, "getAllItins").mockResolvedValue([
            {
                id: 1,
                username: "testUser",
                title: "testItin1",
                city: "testCity1",
                country: "testCountry1",
                duration: 1, 
                description: "Itin for testing 1."
            },
            {
                id: 2,
                username: "testUser",
                title: "testItin2",
                city: "testCity2",
                country: "testCountry2",
                duration: 1, 
                description: "Itin for testing 2."
            }
        ]);
    });

    // Restore mocks after tests
    afterEach(() => {
        vi.restoreAllMocks();
    });

    test("renders without crashing", () => {
        render(
            <MemoryRouter>
                <UserProvider>
                    <ItinSearch/>
                </UserProvider>
            </MemoryRouter>)
    });

    test("matches snapshot", () => {
        const {asFragment} = render(
            <MemoryRouter>
                <UserProvider>
                    <ItinSearch />
                </UserProvider>
            </MemoryRouter>)
        expect(asFragment()).toMatchSnapshot();
    });

    test("renders itineraries as expected", async () => {
        // mocking functions related to ItinCard
        const mockHasLikedItin = vi.fn().mockReturnValue(false);
        const mockToggleLike = vi.fn();

        const {getByText, asFragment} = render(
            <MemoryRouter>
                <UserProvider hasLikedItin={mockHasLikedItin} toggleLike={mockToggleLike}>
                    <ItinSearch />
                </UserProvider>
            </MemoryRouter>)

        // click search to get itineraries
        fireEvent.click(getByText("Search"));
        
        //check that itineraries display after search
        const itinerary1 = await waitFor(() => getByText("testItin1"));
        const itinerary2 = await waitFor(() => getByText("testItin2"));
  
        expect(itinerary1).toBeInTheDocument();
        expect(itinerary2).toBeInTheDocument();

    })
});