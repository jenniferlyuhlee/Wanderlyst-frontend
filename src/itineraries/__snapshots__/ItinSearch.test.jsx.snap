// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ItinSearch component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container px-5 mb-5 body-cont"
  >
    <h3
      class="text-center pb-3"
    >
      Where will you wander next?
    </h3>
    <form
      class="row"
    >
      <div
        class="col-12 form-floating p-0"
      >
        <input
          class="form-control"
          id="title"
          name="title"
          placeholder="title"
          value=""
        />
        <label
          for="title"
        >
          Title
        </label>
      </div>
      <div
        class="col-6 form-floating p-0"
      >
        <input
          class="form-control"
          id="country"
          name="country"
          placeholder="country"
          value=""
        />
        <label
          for="country"
        >
          Country
        </label>
      </div>
      <div
        class="col-6 form-floating p-0"
      >
        <input
          class="form-control"
          id="duration"
          min="1"
          name="duration"
          placeholder="duration"
          type="number"
          value=""
        />
        <label
          for="duration"
        >
          Max duration (days)
        </label>
      </div>
      <div
        class="col-12 form-floating p-0"
      >
        <select
          class="form-select"
          id="tags"
          multiple=""
          name="tags"
          size="3"
          style="height: 5rem;"
        >
          <option
            value="Foodie"
          >
            Foodie
          </option>
          <option
            value="Explorer"
          >
            Explorer
          </option>
          <option
            value="Shopaholic"
          >
            Shopaholic
          </option>
          <option
            value="Naturalist"
          >
            Naturalist
          </option>
          <option
            value="Daredevil"
          >
            Daredevil
          </option>
          <option
            value="Flâneur"
          >
            Flâneur
          </option>
          <option
            value="Reveller"
          >
            Reveller
          </option>
          <option
            value="Aesthete"
          >
            Aesthete
          </option>
          <option
            value="Relaxationist"
          >
            Relaxationist
          </option>
          <option
            value="Photographer"
          >
            Photographer
          </option>
        </select>
        <label
          for="tags"
        >
          Filter by tags
        </label>
      </div>
      <button
        class="btn rounded-pill btn-lg btn-primary mt-3"
      >
        Search
      </button>
    </form>
  </div>
</DocumentFragment>
`;
